cpp/src/shuffler/chunk.cpp

```diff
@@ -12,24 +12,44 @@
 
 namespace rapidsmpf::shuffler::detail {
 
+Chunk::Event::Event(rmm::cuda_stream_view stream, Communicator::Logger& log) : log_(log) {
+    RAPIDSMPF_CUDA_TRY(cudaEventCreateWithFlags(&event_, cudaEventDisableTiming));
+    RAPIDSMPF_CUDA_TRY(cudaEventRecord(event_, stream));
+}
+
+Chunk::Event::~Event() {
+    if (!is_done()) {
+        log_.warn("Event destroyed before CUDA event completed");
+    }
+    cudaEventDestroy(event_);
+}
+
+[[nodiscard]] bool Chunk::Event::is_done() {
+    if (!done_) {
+        done_ = cudaEventQuery(event_) == cudaSuccess;
```

Comment:
*nit* (maybe?): Add documentation that this query is not thread-safe.

Reply:
I don't really see how this is not thread-safe, yes, we may write/overwrite `done_` but the only side-effect may be that `cudaEventQuery` will be called unnecessarily by the multiple threads, the result should nevertheless be the same. Also I noticed this earlier today as well and attempted to simply `return cudaEventQuery(event_) == cudaSuccess;` and thus bypass `done_` entirely, but that didn't help for https://github.com/rapidsai/rapidsmpf/pull/225#issuecomment-2839899277 .

Reply:
I know what the definition is, I'm asking here specifically, how does that affect the result in any form? If the current thread reads `done_` and it's `false`, even though another thread proceeds to immediately turn it into `true`, the current thread will simply pay an extra `cudaEventQuery(event_)` call, but it will necessarily be also evaluated to `true` which will "re-update" `done_` to `true` and also `return true`. So technically the access to `done_` is not thread-safe, but the result is always correct.

Reply:
As per our previous conversation, I've made changes to ensure thread-safety in https://github.com/rapidsai/rapidsmpf/pull/225/commits/f019d8d244f59e1897d7aa2ff7a82f7c9dd24773.